### ISSUE TITLE
[Bug] Example typescript code doesn't work - Argument of type '{ headless: boolean; }' is not assignable to parameter of type 'LaunchOptions'.   Object literal may only specify known properties, and 'headless' does not exist in type 'LaunchOptions'.

### ISSUE THREAD

**Issue by ajanian** [berstend/puppeteer-extra#426]
I am successfully using puppeteer-extra in my typescript project.  I wanted to set headless: false in my launch according to the sample code.  I looked and something appears wrong with my types.  I have stripped the problem down to a trivial sample project.

**Sample Code**

```typescript
import puppeteer from 'puppeteer-extra';
import StealthPlugin from 'puppeteer-extra-plugin-stealth';

puppeteer.use(StealthPlugin());

(async (): Promise<void> => {
    const browser = await puppeteer.launch({ headless: false });
    await browser.close();
})();
```

**Transpile Error**

```bash
❯ tsc -p .
src/index.ts:7:46 - error TS2345: Argument of type '{ headless: boolean; }' is not assignable to parameter of type 'LaunchOptions'.
  Object literal may only specify known properties, and 'headless' does not exist in type 'LaunchOptions'.

7     const browser = await puppeteer.launch({ headless: false });
                                               ~~~~~~~~~~~~~~~


Found 1 error.
```

**Versions**

```bash
❯ npx envinfo@latest --system --binaries --npmPackages '*(puppeteer*|playwright*|automation-extra*|@extra*)'
Need to install the following packages:
  envinfo@latest
Ok to proceed? (y)

  System:
    OS: macOS 11.2
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 54.80 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 15.8.0 - /usr/local/bin/node
    npm: 7.5.0 - /usr/local/bin/npm
  npmPackages:
    puppeteer-extra-plugin-stealth: ^2.7.4 => 2.7.4
```

**Package.json**

```json
{
  "name": "pex",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "Andrew Janian <andrew@janian.net>",
  "license": "ISC",
  "dependencies": {
    "puppeteer-extra-plugin-stealth": "^2.7.4"
  }
}
```

**Comment by otaviobps** [comment]
I'm having the same issue. It seems this package is not using all possible types from `puppeteer`.

While `puppeteer` uses `options?: (pup.LaunchOptions & pup.ChromeArgOptions & pup.BrowserOptions & {product?: "chrome" | "firefox" | undefined, extraPrefsFirefox?: Record<string, unknown> | undefined}) | undefined)`, this package only uses `options?: pup.LaunchOptions | undefined)`

**Comment by berstend** [comment]
@ajanian your dependencies don't list `puppeteer` or `puppeteer-extra`? 🤔 

In any case, please read this:
https://github.com/berstend/puppeteer-extra/issues/428#issuecomment-778679665

**Comment by ajanian** [comment]
Those dependencies are installed automatically because the plug-in depends on them, right?  It was a trivial example.  In my real code I have them both declared - it doesn't make a difference.

I was able to get it working by changing the launch function signature in puppeteer-extra type locally to have the same arguments as puppeteer.  This is what @otaviobps pointed put as well.

I'm fine if you want to close this because I worked around it but something is definitely wrong with the launch function in the type definition for puppeteer-extra.

**Comment by otaviobps** [comment]
I don't think this issue should be closed. I have both puppeteer and puppeteer-extra in my dependencies. I think this problem comes from missing types in puppeteer-extra launch function. My workaround is using `@ts-ignore`

**Comment by jehy** [comment]
Same here, it looks like a bug.

**Comment by vitorbertolucci** [comment]
It is happening to me to, and it does sound like a bug. I belive the problem is that puppeter-extra's launch function receives the `options` as `Puppeteer.LaunchOptions`, while puppeteer's launch function options is typed as `LaunchOptions & BrowserLaunchArgumentOptions & BrowserConnectOptions &`. The `headless` property comes from `BrowserLaunchArgumentOptions`.

**Comment by Quentin-M** [comment]
Same here, with the following `package.json`
```
    "puppeteer": "^8.0.0",
    "puppeteer-extra": "^3.1.17",
```

**Comment by drogerie21** [comment]
Same here. Typescript Definition LaunchOptions is missing properties like "headless" or "args" and Typescript transpiler is throwing errors.
This Issue shouldn´t be closed!

**Comment by Arch4Arts** [comment]
I get the same error!

```json
    "puppeteer": "^9.1.1",
    "puppeteer-extra": "^3.1.18",
    "puppeteer-extra-plugin-stealth": "^2.7.8",
```

**Comment by Arch4Arts** [comment]
> @ajanian your dependencies don't list `puppeteer` or `puppeteer-extra`? 🤔
> 
> In any case, please read this:
> [#428 (comment)](https://github.com/berstend/puppeteer-extra/issues/428#issuecomment-778679665)

Don't close the problem, you need an investigation.

**Comment by sampaiorafael** [comment]
I'm facing the same error

It seems that there is a deficit in typing to match both libraries. It is impossible to use some stuff on extra, because looks like it aint have the same types declaration.

**Comment by idansof** [comment]
This is caused by a breaking change in Puppeteer 8

![image](https://user-images.githubusercontent.com/20138648/121031510-12001200-c7b3-11eb-8858-ff42289b6762.png)

**Comment by idansof** [comment]
https://github.com/berstend/puppeteer-extra/pull/455

**Comment by arisawali2014** [comment]
is there any update for this?

**Comment by luluhoc** [comment]
any update?

**Comment by vjunjoced** [comment]
any update?

**Comment by gilhrpenner** [comment]
Any update?

**Comment by redareda9** [comment]
Damn, no update. @berstend  Is the project abandoned?

**Comment by wandersonsousa** [comment]
Shit, the way its get away from ts when use puppeteer-extra

**Comment by AntonGsv** [comment]
@berstend the problem is still relevant.

**Comment by lexagr** [comment]
node_modules\puppeteer-extra\dist\index.d.ts:89

> launch(options?: Puppeteer.LaunchOptions): Promise<Puppeteer.Browser>;

change this line to
> launch(options?: Puppeteer.LaunchOptions & Puppeteer.BrowserLaunchArgumentOptions & Puppeteer.BrowserConnectOptions ): Promise<Puppeteer.Browser>;

**Comment by nrodriguez** [comment]
Anyone heard of any timeline to fix this issue?

**Comment by berstend** [comment]
> Anyone heard of any timeline to fix this issue?

This is now a high-priority issue and actively being worked on.

---

Long form answer:

We've been waiting a bit for the dust to settle on the unfortunate recurring breaking changes of typings in puppeteer but it's about time to implement a solution.

Quick recap why this is a little tricky:

In [puppeteer v6](https://github.com/puppeteer/puppeteer/releases/tag/v6.0.0) the project dropped `@types/puppeteer` and shipped with types included. Unfortunately those we're pretty broken/subpar in comparison and I couldn't find a way to tell TS to ignore the built-in types in favor of the external ones. This is now however irrelevant as the external types are out of date by now and the internal ones have improved over time.

In puppeteer v6 through v8 typings were even more wild west but since v9 types seem to have stabilized.

The issue for us: There's no versioned or conditional type exports in TS, so it seems impossible to stay backwards compatible for TS users once we switched to the built-in types. :-/ The fact that we're using TS ourselves and depend on specific upstream types while sitting in between as a plugin framework makes this even harder. 😅 

I'm testing possible workarounds currently but if that doesn't pan out we'll drop type support for older puppeteer versions, which is unfortunate as they're still being used:

![image](https://user-images.githubusercontent.com/1368633/133715403-53deaecc-6512-44d6-969f-b21305620518.png)

**Comment by libid0nes** [comment]
Do I understand correctly that using v10, there is no need to install @ types / puppeteer?

**Comment by berstend** [comment]
> Do I understand correctly that using v10, there is no need to install @types/puppeteer?

Yes, since pptr v6 they themselves switched to using typescript and export the types directly. If `@types/puppeteer` is installed with a recent pptr version it'll simply be ignored

**Comment by berstend** [comment]
I think I found a way to support typings for all puppeteer versions out of the box 😍

**Comment by berstend** [comment]
This should now be fixed through #555:
```
Successfully published:
 - puppeteer-extra-plugin-adblocker@2.12.0
 - puppeteer-extra-plugin-anonymize-ua@2.3.0
 - puppeteer-extra-plugin-block-resources@2.3.0
 - puppeteer-extra-plugin-click-and-wait@2.3.0
 - puppeteer-extra-plugin-devtools@2.3.0
 - puppeteer-extra-plugin-flash@2.3.0
 - puppeteer-extra-plugin-font-size@2.3.0
 - puppeteer-extra-plugin-recaptcha@3.5.0
 - puppeteer-extra-plugin-repl@2.3.0
 - puppeteer-extra-plugin-stealth@2.8.0
 - puppeteer-extra-plugin-user-data-dir@2.3.0
 - puppeteer-extra-plugin-user-preferences@2.3.0
 - puppeteer-extra-plugin@3.2.0
 - puppeteer-extra@3.2.0
```
